File: frontend/src/lib/acuerdo.ts
// Acuerdo entre dos evaluadores (kappa de Cohen sin ponderar). Espejo de
// rosa/acuerdo.py para pintar el conjunto dorado sin pedirlo al servidor.

import type { CasoDorado } from '../datos/tipos';

export function kappaCohen(a: string[], b: string[]): number | null {
  const n = a.length;
  if (n === 0 || b.length !== n) return null;
  const cats = [...new Set([...a, ...b])];
  if (cats.length <= 1) return null;
  let po = 0;
  const fa = new Map<string, number>();
  const fb = new Map<string, number>();
  for (let i = 0; i < n; i++) {
    if (a[i] === b[i]) po++;
    fa.set(a[i]!, (fa.get(a[i]!) ?? 0) + 1);
    fb.set(b[i]!, (fb.get(b[i]!) ?? 0) + 1);
  }
  po /= n;
  let pe = 0;
  for (const c of cats) pe += ((fa.get(c) ?? 0) * (fb.get(c) ?? 0)) / (n * n);
  if (pe >= 1) return null;
  return Math.round(((po - pe) / (1 - pe)) * 10000) / 10000;
}

export function interpretarKappa(k: number | null): string {
  if (k === null) return 'sin datos';
  if (k < 0) return 'peor que el azar';
  if (k >= 0.81) return 'casi perfecto';
  if (k >= 0.61) return 'sustancial';
  if (k >= 0.41) return 'moderado';
  if (k >= 0.21) return 'regular';
  return 'leve';
}

export interface AcuerdoDorado {
  n: number;
  bruto: number | null;
  kappa: number | null;
  interpretacion: string;
}

export function acuerdoDe(casos: CasoDorado[]): AcuerdoDorado {
  const a = casos.map((c) => c.veredictoJuez);
  const b = casos.map((c) => c.veredictoHumano);
  const k = kappaCohen(a, b);
  return { n: casos.length, bruto: casos.length ? Math.round((casos.filter((c) => c.veredictoJuez === c.veredictoHumano).length / casos.length) * 1000) / 1000 : null, kappa: k, interpretacion: interpretarKappa(k) };
}

export function acuerdoPorComprobacion(casos: CasoDorado[]): Record<string, AcuerdoDorado> {
  const salida: Record<string, AcuerdoDorado> = {};
  for (const c of new Set(casos.map((x) => x.comprobacion))) salida[c] = acuerdoDe(casos.filter((x) => x.comprobacion === c));
  return salida;
}

File: rosa/acuerdo.py
"""Acuerdo entre dos evaluadores: kappa de Cohen, kappa ponderado, AC1 de Gwet.

Sirve para saber si el juez (o el Killer entero) acierta frente a una persona
cualificada, por tipo de comprobacion y no como promedio global. Kappa
corrige el acuerdo por el que se esperaria por azar; el ponderado (pesos
lineales) trata las categorias ordenadas (bajo, algunas dudas, alto) de modo
que equivocarse en un escalon cuesta menos que en dos. AC1 de Gwet resiste la
paradoja de kappa cuando una categoria domina (casi todo "acierta"), y se
reporta al lado, no en su lugar. Todo es determinista y sin dependencias.

Referencias: Cohen 1960 y 1968; Landis y Koch 1977 (los tramos de
interpretacion, que ellos mismos llamaron arbitrarios); Gwet 2008.
"""

from __future__ import annotations

import random
from collections import Counter
from typing import Any, Sequence

TRAMOS_LANDIS_KOCH = (
    (0.81, "casi perfecto"),
    (0.61, "sustancial"),
    (0.41, "moderado"),
    (0.21, "regular"),
    (0.0, "leve"),
)


def interpretar(kappa: float | None) -> str:
    if kappa is None:
        return "sin datos"
    if kappa < 0:
        return "peor que el azar"
    for umbral, nombre in TRAMOS_LANDIS_KOCH:
        if kappa >= umbral:
            return nombre
    return "leve"


def matriz_confusion(a: Sequence[Any], b: Sequence[Any], categorias: Sequence[Any] | None = None) -> dict[str, Any]:
    """Filas: evaluador a; columnas: evaluador b."""
    if len(a) != len(b):
        raise ValueError("las dos listas deben tener la misma longitud")
    cats = list(categorias) if categorias is not None else sorted(set(a) | set(b), key=str)
    idx = {c: i for i, c in enumerate(cats)}
    m = [[0] * len(cats) for _ in cats]
    for x, y in zip(a, b):
        m[idx[x]][idx[y]] += 1
    return {"categorias": cats, "matriz": m, "n": len(a)}


def kappa_cohen(a: Sequence[Any], b: Sequence[Any], categorias: Sequence[Any] | None = None, ponderado: bool = False) -> float | None:
    """Kappa de Cohen. Con `ponderado` usa pesos lineales |i - j| / (k - 1)
    sobre el orden de `categorias` (obligatorio darlas en orden). None si no
    hay datos o si el acuerdo esperado es 1 (todo en una categoria)."""
    n = len(a)
    if n == 0 or len(b) != n:
        return None
    mc = matriz_confusion(a, b, categorias)
    cats, m = mc["categorias"], mc["matriz"]
    k = len(cats)
    if k <= 1:
        return None
    filas = [sum(f) for f in m]
    cols = [sum(m[i][j] for i in range(k)) for j in range(k)]
    if ponderado:
        peso = lambda i, j: abs(i - j) / (k - 1)  # noqa: E731
        desacuerdo_obs = sum(peso(i, j) * m[i][j] for i in range(k) for j in range(k)) / n
        desacuerdo_esp = sum(peso(i, j) * filas[i] * cols[j] for i in range(k) for j in range(k)) / (n * n)
        if desacuerdo_esp == 0:
            return None
        return round(1 - desacuerdo_obs / desacuerdo_esp, 4)
    po = sum(m[i][i] for i in range(k)) / n
    pe = sum(filas[i] * cols[i] for i in range(k)) / (n * n)
    if pe >= 1:
        return None
    return round((po - pe) / (1 - pe), 4)


def ac1_gwet(a: Sequence[Any], b: Sequence[Any], categorias: Sequence[Any] | None = None) -> float | None:
    n = len(a)
    if n == 0 or len(b) != n:
        return None
    mc = matriz_confusion(a, b, categorias)
    cats, m = mc["categorias"], mc["matriz"]
    k = len(cats)
    if k <= 1:
        return None
    po = sum(m[i][i] for i in range(k)) / n
    # pi_j: proporcion media de asignaciones a la categoria j entre los dos evaluadores.
    pis = [(sum(m[j]) + sum(m[i][j] for i in range(k))) / (2 * n) for j in range(k)]
    pe = sum(p * (1 - p) for p in pis) / (k - 1)
    if pe >= 1:
        return None
    return round((po - pe) / (1 - pe), 4)


def intervalo_bootstrap(a: Sequence[Any], b: Sequence[Any], categorias: Sequence[Any] | None = None, ponderado: bool = False, repeticiones: int = 1000, semilla: int = 7) -> tuple[float, float] | None:
    """IC 95 % por bootstrap del kappa (percentiles 2,5 y 97,5)."""
    n = len(a)
    if n < 5:
        return None
    rng = random.Random(semilla)
    valores: list[float] = []
    pares = list(zip(a, b))
    for _ in range(repeticiones):
        muestra = [pares[rng.randrange(n)] for _ in range(n)]
        kap = kappa_cohen([x for x, _ in muestra], [y for _, y in muestra], categorias, ponderado)
        if kap is not None:
            valores.append(kap)
    if len(valores) < repeticiones // 2:
        return None
    valores.sort()
    return (round(valores[int(0.025 * len(valores))], 3), round(valores[min(len(valores) - 1, int(0.975 * len(valores)))], 3))


def acuerdo(a: Sequence[Any], b: Sequence[Any], categorias: Sequence[Any] | None = None, ordenadas: bool = False) -> dict[str, Any]:
    """Todo junto, listo para guardar: n, acuerdo bruto, kappa (ponderado si
    las categorias son ordenadas), AC1, IC bootstrap, matriz e interpretacion."""
    n = len(a)
    if n == 0:
        return {"n": 0, "bruto": None, "kappa": None, "ac1": None, "ic95": None, "matriz": None, "interpretacion": "sin datos", "ponderado": ordenadas}
    kap = kappa_cohen(a, b, categorias, ponderado=ordenadas)
    return {
        "n": n,
        "bruto": round(sum(1 for x, y in zip(a, b) if x == y) / n, 3),
        "kappa": kap,
        "ac1": ac1_gwet(a, b, categorias),
        "ic95": intervalo_bootstrap(a, b, categorias, ponderado=ordenadas),
        "matriz": matriz_confusion(a, b, categorias),
        "interpretacion": interpretar(kap),
        "ponderado": ordenadas,
        "distribucion": dict(Counter(str(x) for x in b)),
    }
